### PR TITLE
TRestRawBiPoAnalysisProcess first implementation

### PR DESCRIPTION
**Pull request by Anakintana**
Anakintana Medium: 175 [](https://gitlab.cern.ch/rest-for-physics/rawlib/-/commits/aquintana-BiPo) [](https://github.com/rest-for-physics/rawlib/commits/aquintana-BiPo) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=rest-for-physics&utm_campaign=badge_info)

First attend at implementing TRestRawBiPoAnalysisProcess to get T1-T2 distance (from TRestRawBiPoToSignalProcess) as and observable

### CODE REVIEW

File: src/TRestRawSignalEvent.cxx [hunk 1]
```diff
@@ -354,6 +354,15 @@ Double_t TRestRawSignalEvent::GetMaxTime() {
...
 }
 
+// Specific variable to recover the T2-T1 distance from TRestRawBiPoToSignalProcess
+// to then be used in the new TRestRawBiPoAnalysisProcess
+
+Double_t fAuxiliar = 0;
+
+void SetAuxiliar(Double_t aux) { fAuxiliar = aux; }
+
+auto GetAuxiliar() { return fAuxiliar; }
```
Comment (jgalan) [inline src/TRestRawSignalEvent.cxx:364]: Better move to the header file `TRestRawSignalEvent.h`. It should be placed in the `public` region.

File: src/TRestRawSignalEvent.cxx [hunk 2]
```diff
@@ -354,6 +354,15 @@ Double_t TRestRawSignalEvent::GetMaxTime() {
     return maxTime;
 }
 
+// Specific variable to recover the T2-T1 distance from TRestRawBiPoToSignalProcess
```
Comment (jgalan) [inline src/TRestRawSignalEvent.cxx:357]: The comment should be one line long. I would better write something more generic.

`/// An auxiliar variable that can be used to store temporal values.

**Comment by jgalan** [comment]
Some methods are missing inside `TRestRawBiPoAnalysisProcess`.

You may use the macro `REST_MakeProcess.C` in order to create an skeleton template for your process.

You can do:

```
restRootMacros
REST_MakeProcess("TRestRawBiPoAnalysisProcess", "TRestRawSignalEvent", "TRestRawSignalEvent");
```

File: src/TRestRawBiPoAnalysisProcess.cxx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+/// \brief The main processing event function
+///
+TRestEvent* TRestRawBiPoAnalysisProcess::ProcessEvent(TRestEvent* evInput) {
+    fSignalEvent = (TRestRawSignalEvent*)evInput;
+
+    // Write here the main logic of process: TRestRawBiPoAnalysisProcess
+    // Read data from input event, write data to output event, and save observables to tree
+
+    Double_t t1t2_BiPo = event.GetAuxiliar();
+    SetObservableValue("T1-T2_distance", t1t2_BiPo);
```
Comment (jgalan) [inline src/TRestRawBiPoAnalysisProcess.cxx:92]:
```suggestion
    SetObservableValue("t1t2", t1t2_BiPo);
```

File: inc/TRestRawBiPoAnalysisProcess.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        // Write here how to print the added process members and parameters.
+
+        EndPrintProcess();
+    }
+
+    TRestRawBiPoAnalysisProcess();
+    ~TRestRawBiPoAnalysisProcess();
+
+    // ROOT class definition helper. Increase the number in it every time
```
Comment (jgalan) [inline inc/TRestRawBiPoAnalysisProcess.h:72]: This comment should be removed. It is just for the template as a user guidance.

File: inc/TRestRawBiPoAnalysisProcess.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    TRestRawSignalEvent* fAnaEvent;  //!
+
+    void Initialize() override;
+
+    // Add here the members or parameters for your event process.
+    // You can set their default values here together.
+    // Note: add "//!" mark at the end of the member definition
+    // if you don't want to save them to disk.
+
+    /// REMOVE THIS MEMBER! A dummy member that will be written to the ROOT file.
```
Comment (jgalan) [inline inc/TRestRawBiPoAnalysisProcess.h:42]: These data members should be removed, are just template documentation.

File: inc/TRestRawBiPoAnalysisProcess.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+/// TODO Write here a brief description. Just one line!
+class TRestRawBiPoAnalysisProcess : public TRestEventProcess {
+   private:
+    /// A pointer to the specific TRestRawSignalEvent input event
+    TRestRawSignalEvent* fAnaEvent;  //!
+
+    void Initialize() override;
+
+    // Add here the members or parameters for your event process.
```
Comment (jgalan) [inline inc/TRestRawBiPoAnalysisProcess.h:37]: Remove also this comment.

File: inc/TRestRawSignalEvent.h [hunk 7]
```diff
@@ -44,6 +44,7 @@ class TRestRawSignalEvent : public TRestEvent {
     Double_t fMaxTime;   //!
     Double_t fMinValue;  //!
     Double_t fMaxValue;  //!
+    // Double_t fAuxiliar = 0;
```
Comment (jgalan) [inline inc/TRestRawSignalEvent.h:47]: Should be uncommented

Comment (jgalan) [inline inc/TRestRawSignalEvent.h:47]:
```suggestion
    Double_t fAuxiliar = 0;
```

File: inc/TRestRawSignalEvent.h [hunk 8]
```diff
@@ -72,6 +73,10 @@ class TRestRawSignalEvent : public TRestEvent {
         for (int n = 0; n < GetNumberOfSignals(); n++) fSignal[n].SetTailPoints(p);
     }
 
+    // Set and Get a tmp variable (for BiPo)
+    void SetAuxiliar(Double_t aux);
+    // auto GetAuxiliar( ) { return fAuxiliar; }
```
Comment (jgalan) [inline inc/TRestRawSignalEvent.h:78]:
```suggestion
    auto GetAuxiliar( ) { return fAuxiliar; }
```

File: src/TRestRawBiPoAnalysisProcess.cxx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+ * You should have a copy of the GNU General Public License along with   *
+ * REST in $REST_PATH/LICENSE.                                           *
+ * If not, see http://www.gnu.org/licenses/.                             *
+ * For the list of contributors see $REST_PATH/CREDITS.                  *
+ *************************************************************************/
+
+////////////////////////////////////////////////////////////////////////
+/// The TRestRawBiPoSignalAnalysisProcess is meant to add specific BiPo
+/// observables when doing the analysis.
+/// For the moment it gives the observable T1-T2_distance
```
Comment (jgalan) [inline src/TRestRawBiPoAnalysisProcess.cxx:26]:
```suggestion
/// For the moment it gives the observable `t1t2` which is the ... (better add a list of observables, even if there is only one).
```

File: src/TRestRawSignalEvent.cxx (hunk 1 above)
Comment (jgalan) [reply src/TRestRawSignalEvent.cxx:360]: These methods should be removed

File: src/TRestRawBiPoAnalysisProcess.cxx [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+///////////////////////////////////////////////
+/// \brief The main processing event function
+///
+TRestEvent* TRestRawBiPoAnalysisProcess::ProcessEvent(TRestEvent* evInput) {
+    fSignalEvent = (TRestRawSignalEvent*)evInput;
+
+    // Write here the main logic of process: TRestRawBiPoAnalysisProcess
+    // Read data from input event, write data to output event, and save observables to tree
+
+    Double_t t1t2_BiPo = event.GetAuxiliar();
```
Comment (jgalan) [inline src/TRestRawBiPoAnalysisProcess.cxx:91]: The pointer is `fSignalEvent->` and not `event.`.

**Comment by Anakintana** [comment]
> The file `TRestRawBiPoToSignalProcess` should be removed now, since there is a new process filename.
> 
> ```
> git rm src/TRestRawBiPoToSignalProcess.cxx
> ```

**Comment by Anakintana** [comment]
> The file `TRestRawBiPoToSignalProcess` should be removed now, since there is a new process filename.
> 
> ```
> git rm src/TRestRawBiPoToSignalProcess.cxx
> ```

Why should it be removed? Don't we need these two processes? (TRestRawBiPoToSignalProcess and TRestRawBiPoAnalysisProcess)

File: src/TRestRawSignalEvent.cxx (hunk 1 above)
Comment (jgalan) [reply src/TRestRawSignalEvent.cxx:360]: I think you still need to remove `GetAuxiliar` here. It is replicated in the header. `fAuxiliar`and `GetAuxiliar`should be removed if they are already in the header.

File: inc/TRestRawBiPoAnalysisProcess.h (hunk 4 above)
Comment (jgalan) [reply inc/TRestRawBiPoAnalysisProcess.h:60]: Please, remove the comment that is coming from the template

**Review by jgalan — APPROVED** [review]

**Comment by Anakintana** [comment]
The checks of this PR get stuck with PandaX validation and we don't understand why this is happening. Maybe @lobis or @juanangp know what's going on.

**Comment by lobis** [comment]
> The checks of this PR get stuck with PandaX validation and we don't understand why this is happening. Maybe @lobis or @juanangp know what's going on.

No idea, probably it just got stuck (sometimes it happens). Just rerunning the ci usually fixes things.

**Comment by lobis** [comment]
> > I tried re-running, but no changes. It is something I had never seen before. @lobis I thought you were discussing about removing the PandaX-III pipelines?
> 
> My guess is because PandaX-III has been already removed? Perhaps github is not looking at the proper ref for the framework-validation pipeline...

https://github.com/rest-for-physics/framework/pull/475

it's been removed yes.

**Comment by jgalan** [comment]
Ok, so, how do we fix the problem? Do I just merge without waiting the requirements to meet? Will it not happen again?

**Pull request merged**